dev/pl011/uart.c

```diff
@@ -83,37 +81,38 @@ static enum handler_return uart_irq(void *arg) {
     return resched ? INT_RESCHEDULE : INT_NO_RESCHEDULE;
 }
 
-void uart_init(void) {
-    for (size_t i = 0; i < NUM_UART; i++) {
-        uintptr_t base = uart_to_ptr(i);
+void pl011_uart_init(int irq, int nr, uintptr_t base) {
+    assert(nr < NUM_UART);
+    uart_base[nr] = base;
+    // create circular buffer to hold received data
+    cbuf_initialize(&uart_rx_buf[nr], RXBUF_SIZE);
 
-        // create circular buffer to hold received data
-        cbuf_initialize(&uart_rx_buf[i], RXBUF_SIZE);
+    // assumes interrupts are contiguous
+    register_int_handler(irq, &uart_irq, (void *)nr);
```

Comment:
the (void *) is a warning on arm64. Perhaps as before, cast it through uintptr_t first.